rust/foxglove/src/websocket.rs

```diff
@@ -470,6 +544,157 @@ impl ConnectedClient {
+            // This is the behavior of the reference Python server implementation.
+            if request_id.is_some() {
+                let message = protocol::server::parameters_json(&updated_parameters, request_id);
+                self.send_control_msg(Message::text(message));
+            }
+            updated_parameters
+        } else {
+            // This differs from the Python reference implementation in that here we notify
+            // subscribers about the parameters even if there's no ServerListener configured.
+            // This seems to be a more sensible default.
+            parameters
+        };
+        server.publish_parameter_values(updated_parameters);
+    }
+
+    fn update_parameters(&self, parameters: &[Parameter]) {
+        // Hold the lock for as short a time as possible
+        let subscribed_parameters: Vec<Parameter> = {
+            let subscribed_parameters = self.parameter_subscriptions.lock();
+            // Filter parameters to only send the ones the client is subscribed to
+            parameters
+                .iter()
+                .filter(|p| subscribed_parameters.contains(&p.name))
+                .cloned()
+                .collect()
+        };
+        if subscribed_parameters.is_empty() {
+            return;
+        }
+        let message = protocol::server::parameters_json(&subscribed_parameters, None);
+        self.send_control_msg(Message::text(message));
+    }
+
+    fn on_parameters_subscribe(&self, server: Arc<Server>, param_names: Vec<String>) {
+        if !server
+            .capabilities
+            .contains(&Capability::ParametersSubscribe)
+        {
+            self.send_error("Server does not support parametersSubscribe capability".to_string());
+            return;
+        }
+
+        // We hold the server lock here the entire time to serialize
+        // calls to subscribe and unsubscribe, otherwise there are all
+        // kinds of race conditions here where handlers get invoked in
+        // an order different than the order the events were applied,
+        // leading to the listener thinking it has no subscribers to a
+        // parameter when it actually does or visa versa.
+        let mut new_param_subscriptions = Vec::with_capacity(
+            self.server_listener
+                .as_ref()
+                .map(|_| param_names.len())
+                .unwrap_or_default(),
+        );
+        let mut all_subscriptions = server.subscribed_parameters.lock();
+
+        // Get the list of which subscriptions are new to the server (first time subscriptions)
+        if self.server_listener.is_some() {
+            for name in param_names.iter() {
+                if all_subscriptions.insert(name.clone()) {
+                    new_param_subscriptions.push(name.clone());
+                }
+            }
+        }
+
+        // Track the client's own subscriptions
+        let mut client_subscriptions = self.parameter_subscriptions.lock();
```

Comment:
I see why the lock on server subscriptions is held across the callback, but why this one too?

Reply:
Yeah, good catch